### PR TITLE
feat: use basepath name for project name and add on-create hook

### PR DESCRIPTION
**Pull request by sor4chi**
Hi, @yusukebe 

I implemented what we discussed in https://github.com/honojs/hono/issues/1563 to try it out. What do you think?

### CODE REVIEW

File: src/hooks/on-create.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { readFileSync, writeFileSync } from 'fs'
+import path from 'path'
+
+type Hook = (args: { projectName: string; directoryPath: string }) => void
```
Comment (yusukebe) [inline src/hooks/on-create.ts:4]: This type definition should not be placed on this file. It's good to create `hooks/types.ts`.

File: src/hooks/on-create.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+// <templateName, Hook[]>
+export const ON_CREATE_HOOKS = new Map<string, Hook[]>()
```
Comment (yusukebe) [inline src/hooks/on-create.ts:7]: I think `AFTER_CREATE` is better than `ON_CREATE_HOOKS`.

File: src/hooks/on-create.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export const addOnCreateHook = (templateName: string, hook: Hook) => {
+  const hooks = ON_CREATE_HOOKS.get(templateName) || []
+  hooks.push(hook)
+  ON_CREATE_HOOKS.set(templateName, hooks)
+}
+
+const PROJECT_NAME_REPLACE_KEY = '[DYNAMIC_PROJECT_NAME]'
```
Comment (yusukebe) [inline src/hooks/on-create.ts:15]: I like the usage of `%%PROJECT_NAME%%` as it is used in general template engines.

File: src/index.ts [hunk 4]
```diff
@@ -45,6 +45,27 @@ async function main() {
   })
   let templateNames = [...Object.values(templates)] as { name: string }[]
 
+  let target = ''
+  let projectName = ''
+  if (args._[0]) {
+    target = String(args._[0])
```
Comment (yusukebe) [inline src/index.ts:51]: The following is better.

```ts
target = args._[0].toString()
```

File: src/index.ts [hunk 5]
```diff
@@ -45,6 +45,27 @@ async function main() {
...
+    projectName = target
+    console.log(`${bold(`${green(`✔`)} Using target directory`)} … ${target}`)
+  } else {
+    const ans = await prompts({
```
Comment (yusukebe) [inline src/index.ts:55]: I like using `answer` instead of `ans`.

**Comment by yusukebe** [comment]
@sor4chi,

Great work! That's exactly what I was looking for.

I've left some comments, so please check them.

And... do you think we can write tests? It would be better to have some, but I understand it might be difficult for this kind of CLI. I know there are some `create-xxx` CLIs for major frameworks that don't have tests.

**Comment by sor4chi** [comment]
Suggested edit:
```diff
diff --git a/src/hooks/on-create.ts b/src/hooks/on-create.ts
index 5468b70..fd14c9e 100644
--- a/src/hooks/on-create.ts
+++ b/src/hooks/on-create.ts
@@ -1,7 +1,6 @@
 import { readFileSync, writeFileSync } from 'fs'
 import path from 'path'
-
-type Hook = (args: { projectName: string; directoryPath: string }) => void
+import { Hook } from './types'
 
 // <templateName, Hook[]>
 export const ON_CREATE_HOOKS = new Map<string, Hook[]>()
diff --git a/src/hooks/types.ts b/src/hooks/types.ts
new file mode 100644
index 0000000..14dfe04
--- /dev/null
+++ b/src/hooks/types.ts
@@ -0,0 +1,4 @@
+export type Hook = (args: {
+  projectName: string
+  directoryPath: string
+}) => void

```

**Comment by sor4chi** [comment]
@yusukebe 

Thank you for the reviews. I resolved the all conversations.

About the tests, If you need me to do it, I'll do it.
But I think it's difficult too...

**Comment by raphaelkieling** [comment]
Hello! 

Sorry for the intrusion, just a free time checking random PRs hehe! What about test only the core logic `after-create.ts` instead of a integration test in the cli itself (since is the hard part)? check if the templates hashmap is being populated correctly, if the hooks are being applied in the correct order and mainly if you are really replacing the correct place in the template file.

Seems important since the quantity of templates should grow with different logics, i can help you if you want. Thanks!

**Comment by yusukebe** [comment]
Hi @raphaelkieling,

No problem! I agree. @sor4chi, can you try it?

I think an integration test is hard because it has to create an actual project (but, it may be possible if we do proper cleanup?).

File: src/hooks/after-create.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { readFileSync, writeFileSync } from 'fs'
+import path from 'path'
+import { Hook } from './types'
+
+// <templateName, Hook[]>src/hooks/on-create.ts
+export const AFTER_CREATE = new Map<string, Hook[]>()
+
+export const addOnCreateHook = (templateName: string, hook: Hook) => {
```
Comment (yusukebe) [inline src/hooks/after-create.ts:8]: I think this should be `addAfterCreateHook`.

**Comment by sor4chi** [comment]
Hi, @raphaelkieling , @yusukebe

I tried to write some tests, how about this?

File: package.json [hunk 7]
```diff
@@ -3,6 +3,7 @@
   "version": "0.2.6",
   "scripts": {
     "build": "tsx ./build.ts",
+    "test": "vitest",
```
Comment (yusukebe) [inline package.json:6]: `vitest --run` is better. We can use this command for CI.

**Review by yusukebe — COMMENTED** [review]
Commented. And please delete `pnpm-lock.yaml` .

**Comment by yusukebe** [comment]
@sor4chi 

What do you think creating a class for the "Hook" like the following:

https://github.com/honojs/create-hono/blob/current-directory-name-class/src/hook.ts

This will be useful to make other hooks and make the code clean, I think.

**Comment by sor4chi** [comment]
@yusukebe 
Cool, it sure looks like it would be easy to expand. 
Surely, my code is redundant because I have to define `addAfterCreateHook` every time.

I will adopt the code as is. Thank you!

**Comment by yusukebe** [comment]
@sor4chi 

Thanks. Use this PR! https://github.com/sor4chi/create-hono/pull/1

**Comment by sor4chi** [comment]
@yusukebe 
By the way, is it ok to make `hookMap` public to test `addHook`?

**Comment by yusukebe** [comment]
I'd prefer not to. I think it's not needed if `addHook` is used in `rewriteWranglerHook` and will be tested accordingly.

**Comment by sor4chi** [comment]
Thank you. I understand!

**Comment by sor4chi** [comment]
Done!

File: src/hooks/after-create.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+type AfterHookFunction = (options: AfterHookOptions) => void
+
+const afterCreateHook = new Hook<AfterHookFunction>()
+
+afterCreateHook.addHook(
+  'cloudflare-workers',
+  ({ projectName, directoryPath }) => {
+    const wranglerPath = path.join(directoryPath, 'wrangler.toml')
+    const wrangler = readFileSync(wranglerPath, 'utf-8')
+    const rewritten = wrangler.replace('%%PROJECT_NAME%%', projectName)
```
Comment (yusukebe) [inline src/hooks/after-create.ts:19]: `replaceAll` is better. Currently there is only one target `name = "%%PROJECT_NAME%%"`, but more may be added in the future.

**Review by yusukebe — COMMENTED** [review]
I've left a comment.

**Comment by sor4chi** [comment]
Opps, i fixed!

**Comment by yusukebe** [comment]
Cool! Looks good!

@raphaelkieling do you have any comment?

**Comment by raphaelkieling** [comment]
@yusukebe @sor4chi awesome work, definitely a more resilient approach since would be really easy to extend for `beforeCreate` or any other necessary hook.

My suggestion for now or in near future is to create file called `cloudflare-workers.hooks.ts` inside the `hooks` folder to avoid put specific template requirements inside the `after-create.ts`. With this, the `wrangler.hooks.ts` can centralize all the logic for `afterCreate` and the `after-create.ts` can centralize ONLY the function interfaces and the `new Hook<...>()`.

What do you think?

**Comment by yusukebe** [comment]
Hi @raphaelkieling,

Thank you for sharing your suggestion! Your way is good for a separation of concerns. I think we should not do it now, but we can consider it in the near future when new hooks will be added. So, I'd like to go with this for now.

@sor4chi, what do you think?

**Comment by sor4chi** [comment]
@yusukebe @raphaelkieling 

I think it's fine to release it once with this and sort it out if it becomes a problem, as I can add it later as well, and I don't think there's much benefit enough to manage it that neatly right now.

**Comment by yusukebe** [comment]
Okay! Let's ship it.

**Pull request merged**